### PR TITLE
[watchOS][targets] Add support for 32bit Apple Watch.

### PR DESCRIPTION
**Pull request by sbogolepov**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt [hunk 1]
```diff
@@ -53,6 +53,9 @@ private class LlvmPipelineConfiguration(context: Context) {
         KonanTarget.IOS_X64 -> "core2"
         KonanTarget.TVOS_ARM64 -> "cyclone"
         KonanTarget.TVOS_X64 -> "core2"
+        KonanTarget.WATCHOS_X86 -> "i386"
+        KonanTarget.WATCHOS_X64 -> "core2"
+        KonanTarget.WATCHOS_ARM32 -> "generic"
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt]: Indeed `generic`?

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt [hunk 2]
```diff
@@ -256,18 +256,30 @@ internal fun Type.isStret(target: KonanTarget): Boolean {
         KonanTarget.MACOS_X64,
         KonanTarget.WATCHOS_X64,
         KonanTarget.TVOS_X64 -> when (unwrappedType) {
-            is RecordType -> unwrappedType.decl.def!!.size > 16 || this.hasUnalignedMembers()
+                is RecordType -> unwrappedType.decl.def!!.size > 16 || this.hasUnalignedMembers()
+                else -> false
+            }
+
+        // See: https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/LowLevelABI/130-IA-32_Function_Calling_Conventions/IA32.html#//apple_ref/doc/uid/TP40002492-SW5
+        // Structures 1 or 2 bytes in size are placed in EAX.
```
Comment (olonho) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: Is it true for structures with float/double members?

Comment (sbogolepov) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: Yep.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExport.kt [hunk 3]
```diff
@@ -154,6 +154,8 @@ internal class ObjCExport(val context: Context, symbolTable: SymbolTable) {
             KonanTarget.TVOS_ARM64 -> "AppleTVOS"
             KonanTarget.TVOS_X64 -> "AppleTVSimulator"
             KonanTarget.MACOS_X64 -> "MacOSX"
+            KonanTarget.WATCHOS_ARM32, KonanTarget.WATCHOS_ARM64 -> "watchOS"
+            KonanTarget.WATCHOS_X86, KonanTarget.WATCHOS_X64 -> "watchSimulator"
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExport.kt]: Xcode generates `watchOS` and `watchSimulator` capitalized here, doesn't it?

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt [hunk 4]
```diff
@@ -256,18 +256,30 @@ internal fun Type.isStret(target: KonanTarget): Boolean {
         KonanTarget.MACOS_X64,
         KonanTarget.WATCHOS_X64,
         KonanTarget.TVOS_X64 -> when (unwrappedType) {
-            is RecordType -> unwrappedType.decl.def!!.size > 16 || this.hasUnalignedMembers()
+                is RecordType -> unwrappedType.decl.def!!.size > 16 || this.hasUnalignedMembers()
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: Indent.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt [hunk 5]
```diff
@@ -256,18 +256,30 @@ internal fun Type.isStret(target: KonanTarget): Boolean {
...
+                val size = unwrappedType.decl.def!!.size
+                val canBePassedInRegisters = (size == 1L || size == 2L || size == 4L || size == 8L)
+                return !canBePassedInRegisters || this.hasUnalignedMembers()
+            }
             else -> false
         }
 
-        KonanTarget.IOS_ARM32 -> when (unwrappedType) {
+        KonanTarget.IOS_ARM32,
+        KonanTarget.WATCHOS_ARM32 -> when (unwrappedType) {
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: This seems wrong.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt [hunk 6]
```diff
@@ -256,18 +256,30 @@ internal fun Type.isStret(target: KonanTarget): Boolean {
...
+
+        // See: https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/LowLevelABI/130-IA-32_Function_Calling_Conventions/IA32.html#//apple_ref/doc/uid/TP40002492-SW5
+        // Structures 1 or 2 bytes in size are placed in EAX.
+        // Structures 4 or 8 bytes in size are placed in: EAX and EDX.
+        // Structures of other sizes are placed at the address supplied by the caller.
+        KonanTarget.WATCHOS_X86 -> when (unwrappedType) {
+            is RecordType -> {
+                val size = unwrappedType.decl.def!!.size
+                val canBePassedInRegisters = (size == 1L || size == 2L || size == 4L || size == 8L)
+                return !canBePassedInRegisters || this.hasUnalignedMembers()
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: This seems wrong:
```c
struct __attribute__((packed)) S {
    float x;
    char y;
};
```

Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: This seems wrong:
```c
struct __attribute__((packed)) S {
    char x;
    short y;
    char z;
};
```

Comment (olonho) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt]: Generally this ad hoc ABI looks like a hack, which we can tolerate now, but likely shall have better solution in the future.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            else -> false
+        }
+    }
+}
+
+class DarwinX86AbiInfo : ObjCAbiInfo {
+    override fun shouldUseStret(returnType: Type): Boolean {
+        // See: https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/LowLevelABI/130-IA-32_Function_Calling_Conventions/IA32.html#//apple_ref/doc/uid/TP40002492-SW5
+        // Structures 1 or 2 bytes in size are placed in EAX.
+        // Structures 4 or 8 bytes in size are placed in: EAX and EDX.
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt]: Not sure if the comment is correct. Isn't `struct S { float x; }` returned on FPU stack?

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    override fun shouldUseStret(returnType: Type): Boolean = when (target) {
+        KonanTarget.IOS_ARM32 -> when (returnType) {
+            is RecordType -> !returnType.isIntegerLikeType()
+            else -> false
+        }
+        // 32-bit watchOS uses armv7k which is effectively Cortex-A7 and
+        // uses AAPCS16 VPF.
+        KonanTarget.WATCHOS_ARM32 -> when (returnType) {
+            is RecordType -> {
+                // https://github.com/llvm/llvm-project/blob/6c8a34ed9b49704bdd60838143047c62ba9f2502/clang/lib/CodeGen/TargetInfo.cpp#L6306
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt]: This link refers code for variadic arguments, not return values.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt (hunk 7 above)
Comment (sbogolepov) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt]: You are correct.
Turns out, this kind of structs is a special case. https://github.com/llvm/llvm-project/blob/6c8a34ed9b49704bdd60838143047c62ba9f2502/clang/lib/CodeGen/TargetInfo.cpp#L1438

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt [hunk 9]
```diff
@@ -0,0 +1,129 @@
+package org.jetbrains.kotlin.native.interop.gen
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt]: Copyright missing.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExport.kt (hunk 3 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExport.kt]: Is this fixed?

**Comment by SvyatoslavScherbina** [comment]
See also: https://github.com/JetBrains/kotlin-native/blob/992afd5aed16074b98ef1ee73836002396d0c6a0/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt#L1253

**Pull request merged**